src/java.base/share/classes/jdk/internal/reflect/MethodAccessorGenerator.java

```diff
@@ -134,6 +135,10 @@ private MagicAccessorImpl generate(final Class<?> declaringClass,
     {
         ByteVector vec = ByteVectorFactory.create();
         asm = new ClassFileAssembler(vec);
+        this.isConstructingPrimitive = isConstructor && declaringClass.isPrimitiveClass();
+        if (this.isConstructingPrimitive ) {
+            returnType = declaringClass.asValueType();
+        }
```

Comment:
Alternatively, this method can take an additional `isStaticFactory` parameter.  `generateConstructor` will determine if the declaring class is a primitive class and then call the static factory with the return type argument rather than `Void.TYPE`.